treewide: Support aarch64-android using minimal prebuilt components

s1341 opened the pull request:
###### Motivation for this change
This PR adds a new aarch64 android toolchain, which leverages the existing crossSystem infrastructure and LLVM builders to generate a working toolchain with minimal prebuilt components.

The only thing that is prebuilt is the bionic libc. This is because it is practically impossible to compile bionic outside of an AOSP tree. I tried and failed, braver souls may prevail. For now I just grab the relevant binaries from https://android.googlesource.com/. 

I also grab the msm kernel sources from there to generate headers. I've included a minor patch to the existing kernel-headers derivation in order to expose an internal function.

Everything else, from binutils up, is using stock code. Many thanks to @Ericson2314 for his help on this, and for building such a powerful system in the first place!

One motivation for this is to be able to build a toolchain which will work on an aarch64 linux machine. To my knowledge, there is no existing toolchain for an aarch64-linux builder and an aarch64-android target.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -19441,6 +19446,8 @@ in
 
   lkl = callPackage ../applications/virtualization/lkl { };
 
+  inherit (callPackages ../os-specific/linux/kernel-headers { })
+    makeLinuxHeaders;
   inherit (callPackages ../os-specific/linux/kernel-headers { })
     linuxHeaders;
```
```suggestion
  inherit (callPackages ../os-specific/linux/kernel-headers { })
    linuxHeaders makeLinuxHeaders;
```

s1341 replied:
Done.

SuperSandro2000 commented on pkgs/os-specific/linux/kernel-headers/default.nix:
```diff
@@ -36,9 +44,14 @@ let
+      make defconfig 
+      make headers_install
+    '' else ''
+      echo =============================================
```
```suggestion
```

s1341 replied:
Done.

SuperSandro2000 commented on pkgs/os-specific/linux/kernel-headers/default.nix:
```diff
@@ -1,26 +1,34 @@
     # `elf-header` is null when libc provides `elf.h`.
-    nativeBuildInputs = [ perl elf-header ];
+    nativeBuildInputs = [ perl elf-header ] 
+    ++ lib.optional stdenvNoCC.hostPlatform.isAndroid [ 
+      buildPackages.buildPackages.bison
+      buildPackages.buildPackages.flex
+      buildPackages.buildPackages.python 
+    ];
```
```suggestion
    ++ lib.optional stdenvNoCC.hostPlatform.isAndroid (with buildPackages.buildPackages; [ 
      bison
      flex
      python
    ]);
```

can we use python3?

s1341 replied:
Done.

SuperSandro2000 commented on pkgs/os-specific/linux/bionic-prebuilt/default.nix:
```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+/**
+ * Set to 1 if this is an NDK, unset otherwise. See
+ * https://android.googlesource.com/platform/bionic/+/master/docs/defines.md.
+ */
+#define __ANDROID_NDK__ 1
+
+/**
+ * Major version of this NDK.
+ *
+ * For example: 16 for r16.
+ */
+#define __NDK_MAJOR__ 22
+
+/**
+ * Minor version of this NDK.
+ *
+ * For example: 0 for r16 and 1 for r16b.
+ */
+#define __NDK_MINOR__ 0
+
+/**
+ * Set to 0 if this is a release build, or 1 for beta 1,
+ * 2 for beta 2, and so on.
+ */
+#define __NDK_BETA__ 0
+
+/**
+ * Build number for this NDK.
+ *
+ * For a local development build of the NDK, this is -1.
+ */
+#define __NDK_BUILD__ 7026061
+
+/**
+ * Set to 1 if this is a canary build, 0 if not.
+ */
+#define __NDK_CANARY__ 0
+END
```
Please convert that into a patch.

Ericson2314 replied:
Yes, and also I believe the `__NDK_MAJOR__` and `__NDK_MINOR__` can be gotten from the `stdenv.hostPlatform.sdkVer`?

s1341 replied:
Those are not the same as the SDK version... they are distinct. I don't believe we use the values anywhere. They just need to be defined.

Ericson2314 replied:
OK

Ericson2314 replied:
Actually, I think it might be still better to put them in Nix (even if they are just `let`-bound with the package and not put in `targetPlatform`) just so people are more likely to remember to update them.

SuperSandro2000 commented on pkgs/os-specific/linux/bionic-prebuilt/default.nix:
```diff
@@ -0,0 +1,137 @@
+      "!defined(BIONIC_IOCTL_NO_SIGNEDNESS_OVERLOAD)" "0"
+  '';
+  installPhase= ''
+    # copy the bionic headers
+    echo copying libc files
```
The comment and echo say different things.

s1341 replied:
Done

SuperSandro2000 commented on pkgs/os-specific/linux/bionic-prebuilt/default.nix:
```diff
@@ -0,0 +1,137 @@
+  installPhase= ''
+    # copy the bionic headers
+    echo copying libc files
+    mkdir -p $out/include
```
```suggestion
    mkdir -p $out/include/support/ $out/include/support/
```

s1341 replied:
Done

SuperSandro2000 commented on pkgs/os-specific/linux/bionic-prebuilt/default.nix:
```diff
@@ -0,0 +1,137 @@
+    cp -vr ${kernelHeaders}/include/*  $out/include/
+    substituteInPlace $out/include/asm/sigcontext.h --replace "#ifdef CONFIG_64BIT" "#if 1"
+    # copy the support headers
+    mkdir -p $out/include/support/
```
```suggestion
```

s1341 replied:
Done

SuperSandro2000 commented on pkgs/os-specific/linux/bionic-prebuilt/default.nix:
```diff
@@ -0,0 +1,137 @@
+    # copy the support headers
+    mkdir -p $out/include/support/
+    cp -vr ${ndk_support_headers}* $out/include/support/
+    mkdir -p $out/include/android
```
```suggestion
```

s1341 replied:
Done.

SuperSandro2000 commented on pkgs/os-specific/linux/bionic-prebuilt/default.nix:
```diff
@@ -0,0 +1,137 @@
+      "!defined(BIONIC_IOCTL_NO_SIGNEDNESS_OVERLOAD)" "0"
+  '';
+  installPhase= ''
+    # copy the bionic headers
+    echo copying libc files
+    mkdir -p $out/include
+    cp -vr libc/include/* $out/include
+    echo copying kernel headers
```
maybe those echos can be removed.

s1341 replied:
Done.

SuperSandro2000 commented on pkgs/os-specific/linux/bionic-prebuilt/default.nix:
```diff
@@ -0,0 +1,137 @@
+    arch = "arm64";
+    version = "5.1";
+    src = fetchurl {
+      url =
+        "https://android.googlesource.com/kernel/msm/+archive/db4bfbca77cf68e60e503f84074b853934bce55f.tar.gz";
```
```suggestion
      url = "https://android.googlesource.com/kernel/msm/+archive/db4bfbca77cf68e60e503f84074b853934bce55f.tar.gz";
```

s1341 replied:
Done.

SuperSandro2000 commented:
Also this should most likely go through staging.

SuperSandro2000 commented:
Also this should most likely go through staging.

s1341 commented:
@SuperSandro2000   Thanks for all your comments. I'll address them next week.

Ericson2314 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9608,7 +9608,8 @@ in
   ### SHELLS
 
-  runtimeShell = "${runtimeShellPackage}${runtimeShellPackage.shellPath}";
+  runtimeShell = if stdenv.targetPlatform.isAndroid then "${androidRuntimeShellPackage}/ash" else "${runtimeShellPackage}${runtimeShellPackage.shellPath}";
```
```suggestion
  runtimeShell = if stdenv.targetPlatform.isAndroid
    then "${androidRuntimeShellPackage}/ash"
    else "${runtimeShellPackage}${runtimeShellPackage.shellPath}";
```

s1341 replied:
Done.

Ericson2314 replied:
It looks like you didn't so reformat it?

Ericson2314 replied:
It also would be much better to leave it as bash if at all possible. This will break a lot off thing.

SuperSandro2000 commented on pkgs/development/compilers/llvm/11/compiler-rt/default.nix:
```diff
@@ -33,20 +35,22 @@ stdenv.mkDerivation rec {
     "-DCOMPILER_RT_OS_DIR=baremetal"
   ];
 
+  CFLAGS = (if !bareMetal then "-I${stdenv.cc.libc.dev}/include" else null);
```
```suggestion
  CFLAGS = (if bareMetal then null else "-I${stdenv.cc.libc.dev}/include");
```

s1341 replied:
Done.

Ericson2314 commented on pkgs/development/compilers/llvm/9/default.nix:
```diff
@@ -125,7 +125,10 @@ let
         echo "-B${targetLlvmLibraries.compiler-rt}/lib" >> $out/nix-support/cc-cflags
         echo "-nostdlib++" >> $out/nix-support/cc-cflags
-      '' + mkExtraBuildCommands cc;
+      '' + mkExtraBuildCommands cc
+         + lib.optionalString stdenv.targetPlatform.isAndroid ''
+           echo "-L${bintools.libc}/lib --no-dependent-libraries" >> $out/nix-support/cc-ldflags
+      '';
```
I don't think you need this anymore?

Ericson2314 commented:
So `bareMetal` is supposed to mean not "we haven't yet built a libc with syscalls" but "we will never build a libc with syscalls": it's about the end goal not the bootstrap order. So those flags should be put back in place. Instead, we'll want to had something like
`haveLibc = stdenv.cc.libc != null` and use that to control whether `"-DCMAKE_C_FLAGS=-nodefaultlibs"` is included.

Ericson2314 commented on pkgs/development/compilers/llvm/11/compiler-rt/default.nix:
```diff
@@ -1,15 +1,17 @@
 
   useLLVM = stdenv.hostPlatform.useLLVM or false;
-  bareMetal = stdenv.hostPlatform.parsed.kernel.name == "none";
+  bareMetal = needBareMetal || (stdenv.hostPlatform.parsed.kernel.name == "none");
```
```suggestion
  bareMetal = stdenv.hostPlatform.parsed.kernel.name == "none";
  haveLibc = stdenv.cc.libc != null;
```

s1341 replied:
Done.

s1341 replied:
Done.

Ericson2314 commented on pkgs/development/compilers/llvm/11/compiler-rt/default.nix:
```diff
@@ -33,20 +35,22 @@ stdenv.mkDerivation rec {
     "-DCMAKE_C_COMPILER_WORKS=ON"
     "-DCMAKE_CXX_COMPILER_WORKS=ON"
     "-DCOMPILER_RT_BAREMETAL_BUILD=ON"
+    "-DCMAKE_C_FLAGS=-nodefaultlibs"
```
```suggestion
```

s1341 replied:
Done

s1341 replied:
Done.

Ericson2314 commented on pkgs/development/compilers/llvm/11/compiler-rt/default.nix:
```diff
@@ -33,20 +35,22 @@ stdenv.mkDerivation rec {
+    "-DCMAKE_C_FLAGS=-nodefaultlibs"
     "-DCMAKE_SIZEOF_VOID_P=${toString (stdenv.hostPlatform.parsed.cpu.bits / 8)}"
   ] ++ lib.optionals (useLLVM) [
     "-DCOMPILER_RT_BUILD_BUILTINS=ON"
```
```suggestion
    "-DCOMPILER_RT_BUILD_BUILTINS=ON"
  ] ++ lib.optionals (useLLVM && !haveLibc) [
```

s1341 replied:
Done.

Ericson2314 replied:
Hmm? This was not done.

s1341 commented:
Ok. Those heads have changed again. I'm going to pin to specific commits. On Sunday.

Ericson2314 commented on pkgs/os-specific/linux/bionic-prebuilt/default.nix:
```diff
@@ -0,0 +1,113 @@
+  };
+
+  kernelHeaders = pkgs.makeLinuxHeaders {
+    arch = "arm64";
```
`linuxArch` would be `arm64` anyways, so why do you need this to be a parameter?

Ericson2314 commented:
I would be much more comfortable making libc++abi static in a separate PR.

Ericson2314 commented on pkgs/development/compilers/llvm/11/compiler-rt/default.nix:
```diff
@@ -33,20 +34,21 @@ stdenv.mkDerivation rec {
     "-DCOMPILER_RT_BUILD_LIBFUZZER=OFF"
     "-DCOMPILER_RT_BUILD_PROFILE=OFF"
-  ] ++ lib.optionals (useLLVM || bareMetal) [
+  ] ++ lib.optionals (bareMetal && !haveLibc) [
```
```suggestion
  ] ++ lib.optionals ((useLLVM || bareMetal) && !haveLibc) [
```

Ericson2314 commented on pkgs/os-specific/linux/kernel-headers/default.nix:
```diff
@@ -1,26 +1,35 @@
+  makeLinuxHeaders = { arch, src, version, patches ? [] }: stdenvNoCC.mkDerivation {
     inherit src;
 
+    inherit arch;
```
I don't think you need this?

Ericson2314 commented on pkgs/os-specific/linux/kernel-headers/default.nix:
```diff
@@ -1,26 +1,35 @@
 
     hardeningDisable = lib.optional stdenvNoCC.buildPlatform.isDarwin "format";
 
+    sourceRoot = lib.optionalString stdenvNoCC.hostPlatform.isAndroid ".";
```
What is this for? I would have thought this is the default?

s1341 replied:
The kernel we get from the repo is not in a top level directory. So when we unpack we get all the directories that would normally be under that top level directory instead. Thus we need to set sourceRoot to the current directory.

Ericson2314 replied:
Ok

Ericson2314 commented on pkgs/os-specific/linux/kernel-headers/default.nix:
```diff
@@ -36,9 +45,12 @@ let
       make mrproper $makeFlags
-    '' + ''
+    '' + (if stdenvNoCC.hostPlatform.isAndroid then ''
+      make defconfig
+      make headers_install
```
Do we know why this became needed?

s1341 replied:
I don't know why it is necessary, but it doesn't work without it. Perhaps there is something different in the android fork of the kernel?

Ericson2314 commented:
> Ok. Those heads have changed again. I'm going to pin to specific commits. On Sunday.

Yes please do, or this will keep on breaking.

Commit:
treewide: Support aarch64-android using minimal prebuilt components

This PR adds a new aarch64 android toolchain, which leverages the
existing crossSystem infrastructure and LLVM builders to generate a
working toolchain with minimal prebuilt components.

The only thing that is prebuilt is the bionic libc. This is because it
is practically impossible to compile bionic outside of an AOSP tree. I
tried and failed, braver souls may prevail. For now I just grab the
relevant binaries from https://android.googlesource.com/.

I also grab the msm kernel sources from there to generate headers. I've
included a minor patch to the existing kernel-headers derivation in
order to expose an internal function.

Everything else, from binutils up, is using stock code. Many thanks to
@Ericson2314 for his help on this, and for building such a powerful
system in the first place!

One motivation for this is to be able to build a toolchain which will
work on an aarch64 linux machine. To my knowledge, there is no existing
toolchain for an aarch64-linux builder and an aarch64-android target.

s1341 commented:
Cleaned things up, resulting in the minimum needed for building the working toolchain. Rebased on staging. Moved to llvm 12.

This should hopefully be ready to merge.

Ericson2314 commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -372,8 +372,6 @@ stdenv.mkDerivation {
     + optionalString (libcxx.isLLVM or false) (''
       echo "-isystem ${lib.getDev libcxx}/include/c++/v1" >> $out/nix-support/libcxx-cxxflags
       echo "-stdlib=libc++" >> $out/nix-support/libcxx-ldflags
-    '' + lib.optionalString stdenv.targetPlatform.isLinux ''
```
I don't think we always want to remove this, since you are no longer doing the static linking change in this PR?

s1341 replied:
Right. Good catch.

Ericson2314 commented on pkgs/development/compilers/llvm/12/compiler-rt/default.nix:
```diff
@@ -30,22 +31,24 @@ stdenv.mkDerivation rec {
     "-DDARWIN_osx_ARCHS=${stdenv.hostPlatform.darwinArch}"
     "-DDARWIN_osx_BUILTIN_ARCHS=${stdenv.hostPlatform.darwinArch}"
+  ] ++ lib.optionals (stdenv.hostPlatform.isAndroid) [
+    "-DLLVM_ENABLE_PER_TARGET_RUNTIME_DIR=ON"
```
What made you add this?

s1341 replied:
clang was looking in the `linux` subdir for the builtins. Adding this made it work. Didn't investigate too deeply why it was not working previously.

Ericson2314 replied:
This I broke but have fixed in my last PR #122399.

s1341 replied:
Uh. Ok. So how do I rebase ontop of that PR to test?

s1341 replied:
OK. Removed.

Ericson2314 commented on pkgs/development/compilers/llvm/12/default.nix:
```diff
@@ -131,6 +131,7 @@ let
         echo "-B${targetLlvmLibraries.compiler-rt}/lib" >> $out/nix-support/cc-cflags
       '' + lib.optionalString (!stdenv.targetPlatform.isWasm) ''
         echo "--unwindlib=libunwind" >> $out/nix-support/cc-cflags
+        echo "-lunwind" >> $out/nix-support/cc-ldflags
```
The previous line wasn't enough?

s1341 replied:
Nope.

Ericson2314 replied:
Maybe we should make this only Android and `useLLVM`.

s1341 replied:
Ok.

Ericson2314 commented on pkgs/development/compilers/llvm/12/default.nix:
```diff
@@ -181,13 +182,28 @@ let
+        stdenv = overrideCC stdenv buildLlvmTools.lldClangNoCompilerRtWithLibc;
+      }));
+
+    compiler-rt-nolibc = callPackage ./compiler-rt ({ inherit llvm_meta; } //
```
I would have `compiler-rt` and `compiler-rt-libc`.

s1341 replied:
Ok.

Ericson2314 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9907,7 +9907,10 @@ in
   ### SHELLS
 
-  runtimeShell = "${runtimeShellPackage}${runtimeShellPackage.shellPath}";
+  runtimeShell = if stdenv.targetPlatform.isAndroid
```
Can we skip this change for now?

s1341 replied:
Sure.

Ericson2314 reviewed (commented):
Just a few things, we're getting close!

Ericson2314 commented:
@thefloweringash I think these compiler-rt changes are very similar to what you've done; I'd love to see the implementations deduplicated!
